src/Container/Container.php

```diff
@@ -36,76 +41,121 @@ class Container implements ContainerInterface
      * @param mixed  $value
      * @return Container
      */
-    public function set($key, $value)
+    public function set(string $key, $value): self
     {
+        if (!(\is_string($value) || \is_object($value))) {
+            throw new \InvalidArgumentException('Value must be a class name, an object instance, or a callable');
+        }
+
         $this->dicValues[$key] = $value;
         unset($this->dicObjects[$key]); // In case an object instance was stored for sharing
 
         return $this;
     }
 
+    public function alias(string $alias, string $key): self
+    {
+        $this->aliases[$alias] = $key;
+
+        return $this;
+    }
+
     /**
      * Check if a DIC value/object exists.
      *
      * @param string $key
      * @return bool
      */
-    public function has($key)
+    public function has($key): bool
     {
+        if (array_key_exists($key, $this->aliases)) {
+            $key = $this->aliases[$key];
+        }
+
         return array_key_exists($key, $this->dicValues);
     }
 
     /**
      * @param string $key
      * @return bool
      */
-    public function hasInstance($key)
+    public function hasInstance($key): bool
     {
+        if (array_key_exists($key, $this->aliases)) {
+            $key = $this->aliases[$key];
+        }
+
         return isset($this->dicObjects[$key]);
     }
 
     /**
-     * Get the shared instance of a DIC object, or a DIC value if it's not an object.
+     * Get the shared instance of a DIC object
      *
      * @param string $key
      * @return mixed
      */
     public function get($key)
     {
-        if (!$this->has($key)) {
-            throw new NotFoundException(sprintf('No application value with key "%s" is defined.', $key));
+        if (array_key_exists($key, $this->aliases)) {
+            $key = $this->aliases[$key];
         }
 
         // Get already instantiated object if it exist
         if (isset($this->dicObjects[$key])) {
             return $this->dicObjects[$key];
         }
 
-        // Check if it's an invokable (closure/anonymous function)
-        if (is_object($this->dicValues[$key]) && method_exists($this->dicValues[$key], '__invoke')) {
-            $this->dicObjects[$key] = $this->dicValues[$key]($this);
-
-            return $this->dicObjects[$key];
+        try {
+            if (array_key_exists($key, $this->dicValues)) {
+                if (\is_object($this->dicValues[$key])) {
+                    // Is it an invokable? (closure/anonymous function)
+                    if (method_exists($this->dicValues[$key], '__invoke')) {
+                        $instance = $this->dicValues[$key]($this);
+                    } else {
+                        $instance =  $this->dicValues[$key];
+                    }
+                } else {
+                    $instance = $this->resolveInstance($this->dicValues[$key]);
```

Comment:
If the instance is already in `$this->dicValues` - why can't we return it right away instead of calling `$this->resolveInstance()`?

Reply:
Because instances are actually stored in `$this->dicObjects`, from which we do indeed return an object immediately if one exists (on line 105).